src/graph/pickle.cc

```diff
@@ -14,85 +17,136 @@ namespace dgl {
 
 HeteroPickleStates HeteroPickle(HeteroGraphPtr graph) {
   HeteroPickleStates states;
-  states.metagraph = graph->meta_graph();
-  states.num_nodes_per_type = graph->NumVerticesPerType();
-  states.adjs.resize(graph->NumEdgeTypes());
+  dmlc::MemoryStringStream ofs(&states.meta);
+  dmlc::Stream *strm = &ofs;
+  strm->Write(ImmutableGraph::ToImmutable(graph->meta_graph()));
+  strm->Write(graph->NumVerticesPerType());
+  std::vector<SparseFormat> fmts;
+  std::vector<char> flags;
   for (dgl_type_t etype = 0; etype < graph->NumEdgeTypes(); ++etype) {
     SparseFormat fmt = graph->SelectFormat(etype, SparseFormat::kAny);
-    states.adjs[etype] = std::make_shared<SparseMatrix>();
     switch (fmt) {
-      case SparseFormat::kCOO:
-        *states.adjs[etype] = graph->GetCOOMatrix(etype).ToSparseMatrix();
+      case SparseFormat::kCOO: {
+        fmts.push_back(SparseFormat::kCOO);
+        const auto &coo = graph->GetCOOMatrix(etype);
+        flags.push_back(coo.row_sorted);
+        flags.push_back(coo.col_sorted);
+        states.arrays.push_back(coo.row);
+        states.arrays.push_back(coo.col);
         break;
+      }
       case SparseFormat::kCSR:
-      case SparseFormat::kCSC:
-        *states.adjs[etype] = graph->GetCSRMatrix(etype).ToSparseMatrix();
+      case SparseFormat::kCSC: {
+        fmts.push_back(SparseFormat::kCSR);
+        const auto &csr = graph->GetCSRMatrix(etype);
+        flags.push_back(csr.sorted);
+        states.arrays.push_back(csr.indptr);
+        states.arrays.push_back(csr.indices);
+        states.arrays.push_back(csr.data);
         break;
+      }
       default:
         LOG(FATAL) << "Unsupported sparse format.";
     }
   }
+  strm->Write(fmts);
+  strm->Write(flags);
   return states;
 }
 
 HeteroGraphPtr HeteroUnpickle(const HeteroPickleStates& states) {
-  const auto metagraph = states.metagraph;
-  const auto &num_nodes_per_type = states.num_nodes_per_type;
-  CHECK_EQ(states.adjs.size(), metagraph->NumEdges());
+  char *buf = const_cast<char *>(states.meta.c_str());  // a readonly stream?
```

Comment:
Why not still using `MemoryStringStream`?

Reply:
This line is to drop the `const`. It is still necessary if using `MemoryStringStream`?

Reply:
You can directly use `dmlc::MemoryStringStream ifs(&states.meta);` as how you write.